File: Priority_Queue/1675. Minimize Deviation in Array/1675. Minimize Deviation in Array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        multiset<int>Set;
        
        for (auto x : nums) {
            if (x%2 == 1) {
                Set.insert(x*2);
            } else {
                Set.insert(x);
            }
        }
        
        int rets = INT_MAX;
        while (true) {
            rets = min(rets, *Set.rbegin()-*Set.begin());
            int m = *Set.rbegin();
            *Set.erase(Set.find(m));
            if (m%2 ==0) {
                Set.insert(m/2);
            } else {
                break;
            }
        }
        return rets;
    }
};

File: Priority_Queue/1675. Minimize Deviation in Array/1675. Minimize Deviation in Array.java
class Solution {
    // with 632
    public int minimumDeviation(int[] nums) {
        PriorityQueue<Integer> pq = new PriorityQueue<>( (a, b) -> (b-a) );
        
        int curmin = Integer.MAX_VALUE;
        for (int n : nums) {
            if (n%2 == 1) {
                curmin = Math.min(curmin, n*2);
                pq.offer(n*2);
            } else {
                curmin = Math.min(curmin, n);
                pq.offer(n);
            }
        }
        
        int rets = Integer.MAX_VALUE;
        while (true) {
            rets = Math.min(rets, pq.peek()-curmin);
            
            int cur = pq.poll();
            if (cur%2 ==0) {
                curmin = Math.min(curmin, cur/2);
                pq.offer(cur/2);
            } else {
                break;
            }
        }
        return rets;
    }
}

// candidate array:
// [1, 2]
// [2, 1]
// [3, 6]
// [4, 2, 1]
